ext/opcache/jit/zend_jit_ir.c

```diff
@@ -16685,19 +16763,125 @@ static bool zend_jit_may_be_in_reg(const zend_op_array *op_array, zend_ssa *ssa,
 	return 1;
 }
 
-static void jit_frameless_icall0(zend_jit_ctx *jit, const zend_op *opline)
+static ir_ref jit_frameless_observer(zend_jit_ctx *jit, int checked_stack, const zend_op *opline, ir_ref op1_ref, uint32_t op1_info, ir_ref op2_ref, uint32_t op2_info, ir_ref op1_data_ref, uint32_t op1_data_info, ir_ref res_ref) {
+	// JIT: zend_observer_handler_is_unobserved(ZEND_OBSERVER_DATA(fbc))
+	ir_ref observer_handler;
+	zend_function *fbc = ZEND_FLF_FUNC(opline);
+	// Not need for runtime cache or generator checks here, we just need if_unobserved
+	ir_ref if_unobserved = jit_observer_fcall_is_unobserved_start(jit, fbc, &observer_handler, IR_UNUSED, IR_UNUSED).if_unobserved;
+
+	// push args to a valid call frame
+	bool track_last_valid_opline = jit->track_last_valid_opline;
+	bool use_last_valid_opline = jit->use_last_valid_opline;
+	const zend_op *last_valid_opline = jit->last_valid_opline;
+	bool reuse_ip = jit->reuse_ip;
+	// zend_jit_save_call_chain, but preserve delayed_call_level. We cannot override delayed_call_level without also saving in unobserved path.
+	if (jit->delayed_call_level) {
+		ir_ref call = jit_CALL(jit_FP(jit), call);
+		ir_STORE(jit_CALL(jit_IP(jit), prev_execute_data), jit->delayed_call_level == 1 ? IR_NULL : ir_LOAD_A(call));
+		ir_STORE(call, jit_IP(jit));
+	}
+	zend_jit_push_call_frame(jit, opline, NULL, fbc, 0, 0, checked_stack, ir_CONST_ADDR(fbc), IR_NULL);
+	// yes, we'll temporarily reuse the IP, but we must not handle this via the reuse_ip mechanism - we just reset IP to original in end
+	// the primary issue here is that this code is just executed conditionally while the reuse mechanisms are meant to be global
+	jit->track_last_valid_opline = track_last_valid_opline;
+	jit->use_last_valid_opline = use_last_valid_opline;
+	jit->last_valid_opline = last_valid_opline;
+	jit->reuse_ip = reuse_ip;
```

Comment:
I don't understand why do you save and restore these variables.
You probably need to reset ``jit->last_valid_opline`` to make JIT reload IP when necessary.
``reuse_ip`` should be set if %r15 contains value of ``EX(call)`` and reset if it contains the VM ``opline``.

Reply:
Similar problem to zend_jit_save_call_chain.

Reply:
This won't work. If you keep the old value of ``jit->last_valid_opline``, it's going to be invalid for one of the taken paths. It has to be reset for both paths.

Reply:
That's why we have:
https://github.com/php/php-src/blob/a3ec05ca9d9b59d91e83c754ef81dbbd46d3514c/ext/opcache/jit/zend_jit_ir.c#L16854-L16855

So that it's actually pointing to the same opline for both paths, avoiding any burden on the hot (unobserved) path.

Reply:
OK. Now I understood this. 
It would be definitely better to keep the related logic together and not spread it around functions.